<!DOCTYPE html>
<html>
<head>
<meta charset="utf-8">
<meta name="apple-mobile-web-app-capable" content="yes">

<style>
@import url(CSS/sperm.css);
@import url(CSS/stack.css);
@import url(CSS/d3-tip.css);
@import url(CSS/figures.css);
@import url(CSS/kbroman_talk.css);
/* comment following line to get rid of notes; uncommented to show notes */
/* @import url(CSS/kbroman_presentation.css); */
</style>

<script charset="utf-8" src="JS/d3.js"></script>
<script src="JS/d3.tip.js"></script>
<script src="JS/plotframe.js"></script>

<title>Why aren&rsquo;t all of our graphs interactive?</title>
</head>
<body class="overview">





<section class="titlepage">
<h2 class="highlight">Why aren&rsquo;t all of our graphs
<font id="flash" class="glow">interactive</font>?</h2>

<h4><a href="http://www.biostat.wisc.edu/~kbroman">Karl Broman</a></h4>
<p class="nopadding">Biostatistics &amp; Medical Informatics, University of Wisconsin
&ndash; Madison</p>

<p class="nopadding"><code><a href="http://www.biostat.wisc.edu/~kbroman"
class="gray">www.biostat.wisc.edu/~kbroman</a><br></code></p>

<div style="opacity:0;" id="github">
<p class="nopadding" style="margin-top:-30px;"><code>
<a href="http://github.com/kbroman"
class="gray">github.com/kbroman</a><br>
<a href="http://kbroman.wordpress.com"
class="gray">kbroman.wordpress.com</a><br>
<a class="gray" href="https://twitter.com/kwbroman">@kwbroman</a>
</code></p>
</div>



<script type="text/javascript">
var flashText = d3.select("body").select("#flash");

var flashToggle = 0; // when flashing, keeps track of on/off
var stateToggle = 0; // goes between 0 (off), 1 (on) and 2 (flashing)

var flashFunc =
      function() {
        flashText.attr("class", function() {
          if(flashToggle==5) {
            flashToggle=-1;
            return("highlight"); }
          else return("glow"); });

        flashToggle++;
      };

var flashInterval = setInterval(flashFunc, 400);

flashText.on("click", function() {
    if(stateToggle==0) {
      clearInterval(flashInterval);
      d3.select(this).attr("class", "highlight");
    }
    else if(stateToggle==1) {
      d3.select(this).attr("class", "glow");
    }
    else {
      d3.select(this).attr("class", "highlight");
      flashToggle = 0;
      flashInterval = setInterval(flashFunc, 400);
    }
    stateToggle++;
    if(stateToggle > 2) stateToggle = 0;
});
</script>

<p class="codelink"><a href="presentation.html" class="gray">Remove stickies</a>

<aside>
Click on the link at the bottom-left to go to a version without these
notes.<br><br>

The flashing &ldquo;interactive&rdquo; is a joke; click
on it to switch among off/on/flashing.<br><br>

I've also hidden links to me at github and twitter, just below the
link to my webpage.<br><br>

These slides were tested for Safari on a Mac and might look terrible
in some browsers.  I have a lot more to learn about handling browser
differences.
</aside>

<script language="Javascript">
clicked = 0;
d3.select("div#github")
  .on("mouseover", function() { d3.select(this).style("opacity", 1); })
  .on("mouseout", function() { if(!clicked) d3.select(this).style("opacity", 0); })
  .on("click", function() {
      if(clicked) { clicked=0; d3.select(this).style("opacity", 0); }
      else { clicked=1;  d3.select(this).style("opacity", 1); } });
</script>
</section>





<section>
<h3>Example: <font class=highlight>density estimate</font></h3>
<div id="density_estimate" class="chart"></div>

<div style="font-size:28px;width:420px;align:center;margin-left:350px;margin-top:15px">
  <div style="float:left;" id="bandwidth">Bandwidth = 1.00</div>
  <div id="bgrect" style="float:right;width:135px;">
     <input type="range" min="-2" max="1" step = "0.05" value="0"></div>
</div>
<script src="JS/density.js"></script>

<p class="codelink">
<a href="Coffee/density.coffee" class="gray">code</a> |
<a href="Coffee/plotframe.coffee" class="gray">plotframe</a>

<aside>
In graduate school (at UC Berkeley), I took a course on statistical
computing from <a href="http://www.stat.berkeley.edu/~spector/">Phil
Spector</a>.  One of the projects was to use <a
href="http://homepage.cs.uiowa.edu/~luke/xls/xlispstat/">XLisp-Stat</a>
to create an interactive plot of a density estimate, with a slider for
bandwidth.<br><br>

Why haven&rsquo;t I done anything like this in the
subsequent 15 years?  Rather than make a bunch of static plots to find
a good bandwidth, you can play around with the one interactive plot.
</aside>

<p class="pg">2</p></section>




<section>
<h3>Example: <font class="highlight">scatterplot</font></h3>
<div id="scatterplot" class="chart"></div>
<script src="JS/scatterplot.js"></script>

<p class="codelink">
<a href="Coffee/scatterplot.coffee" class="gray">code</a> |
<a href="Coffee/plotframe.coffee" class="gray">plotframe</a>


<aside>
Outliers can be easily identified, with the right interactive plot.
In R, you can use <code>identify()</code>, but who remembers how to do
it?  Usually, I&rsquo;ll say, &ldquo;Hmm...I want to figure out which
point has weight 1 &gt; 30 and weight 2 &lt; 25.&rdquo;
</aside>

<p class="pg">3</p></section>



<section>

<h3>Opportunities</h3>

<ul class="tight" style="margin-top:-25px;">
  <li>Exploration</li>
       <ul class="lowlight">
         <li>Tuning parameters</li>
         <li>Identifying outliers</li>
         <li>One fancy plot vs 1000 static plots</li>
       </ul>

  <li>Reports for collaborators</li>
       <ul class="lowlight">
         <li>Living documents!</li>
         <li>Allow deeper exploration of the results</li>
         <li>Cut down on simple questions?</li>
       </ul>

  <li><font class="glow">Big Data</font></li>
       <ul class="lowlight">
         <li>Don&rsquo;t just rely on summary statistics</li>
         <li>Greatly compressed information, but with access to the details</li>
         <li>Zoom into dense figures</li>
         <li>More exploration, more connections</li>
       </ul>

  <li>Teaching</li>
       <ul class="lowlight">
         <li>Cool things to look at and play with</li>
         <li>Animated illustrations of key concepts</li>
         <li>Demonstrate data exploration</li>
         <li>Enable intro students to explore data</li>
       </ul>

</ul>

<aside>
The routine use of interactive graphics for exploratory data analysis
seems a bit out of reach, but surely is possible.<br><br>

Incorporating such graphics in reports to our collaborators could be
really valueable.<br><br>

Perhaps the greatest opportunity is for the exploration of large,
complex data sets.<br><br>

Interactive graphs can also make great teaching tools.
</aside>

<p class="pg">4</p></section>



<section>
<h3>Reproducible reports</h3>

<ul class=small>
  <li><a href="http://www.statistik.lmu.de/~leisch/Sweave/">Sweave</a></li>
  <li><a href="http://www.methods.co.nz/asciidoc/">Asciidoc</a>
      <font class="small">[<a href="Examples/clean_genotypes.html" class="gray">Example</a> | <a
      href="Examples/clean_genotypes_asciidoc_source.html" class="gray">Source</a> | <a
      href="http://powerman.name/doc/asciidoc" class="gray">Glossary</a>]</font></li>


  <li><a href="http://daringfireball.net/projects/markdown/">Markdown</a> +
      <a href="http://yihui.name/knitr/">knitr</a> &rarr; <a
  href="http://www.rstudio.org/docs/authoring/using_markdown">R
  markdown</a> (esp. in <a href="http://www.rstudio.org" class="gray">RStudio</a>)</li>
</ul>

<aside>

<a href="http://www.statistik.lmu.de/~leisch/Sweave/">Sweave</a> is
great; I wrote <a href="http://www.rqtl.org/book">my book</a> entirely
with Sweave.  But for routine reports, dealing with page breaks can be
a pain.<br><br>

I&rsquo;ve moved to <a
href="http://www.methods.co.nz/asciidoc/">asciidoc</a> for this;
it&rsquo;s a more readable markup language. I use it to produce an
html report, with no worries about page breaks. (For example, graphs
can be extremely long.)<br><br>

Lots of people are excited about <a
href="http://daringfireball.net/projects/markdown/">markdown</a> with
<a href="http://yihui.name/knitr/">knitr</a>, especially since
it&rsquo;s incorporated within RStudio.  Markdown is like asciidoc
(but even simpler).
</aside>

<p class="pg">5</p></section>


<section>

<h3>Attie project</h3>

<p class="highlight">~500 B6 x BTBR intercross mice<font class="gray">, all knocked out for leptin</font></h4>

<ul class="tight">
  <li>Genotypes at 2057 SNPs (Affymetrix arrays)</li>
  <li>Gene expression in six tissues (Agilent arrays)</li>
      <ul class="lowlight">
        <li>adipose</li>
        <li>gastrocnemius muscle</li>
        <li>hypothalamus</li>
        <li>pancreatic islets</li>
        <li>kidney</li>
        <li>liver</li>
      </ul>
  <li>Numerous clinical phenotypes</li>
      <ul class="nodash"><li class="lowlight">(e.g., body weight, insulin and glucose levels)</li></ul>
  <li>Collaboration with <a
  href="http://www.biochem.wisc.edu/faculty/attie/" class="gray">Alan Attie</a> and colleagues</li>

</ul>



<aside>
My first adventure with big data: 500 mice x 6 arrays x 40k
measurements.  The goal is to identify genes that contribute to
obesity-induced diabetes.<br><br>

With 500 arrays of 40k measurements, you would think we would be
making many more plots than usual.  But, rather, we tend to make <font
class="highlight">no</font> plots<br><br>

If you can&rsquo;t look at all 500 histograms, you tend not to look at any.
</aside>

<p class="pg">5</p></section>



<section>
<h3>Many box plots</h3>


<div align="center">
<img src="Figs/manyboxplots_oldschool.png">
</div>

<aside>500 boxplots, each summarizing 40k measurements.
<br><br> Standard box plots don&rsquo;t work well with large data sets.  The
usual five-number summaries don&rsquo;t give tail information, and it&rsquo;s hard
to show <font class="highlight">many</font> boxplots.</aside>


<p class="pg">6</p></section>

<section>
<h3>Many box plots</h3>


<div align="center">
<img src="Figs/manyboxplots_oldschoolB.png">
</div>

<aside>Boxplots with large datasets are made even worse by use of
the usual rule for depicting &ldquo;outliers.&rdquo; </aside>


<p class="pg">6</p></section>





<section>
<h3>Many box plots, <font class="highlight">modernized</font></h3>


<div align="center">
<img src="Figs/manyboxplots.png">
</div>

<aside>
I like this better: use lines to connect various quantiles. (Here, the
1, 5, 10, 25, 50, 75, 90, 95, and 99%iles.)  <br><br>

Wouldn&rsquo;t it be nice to click on an array and see the corresponding
histogram?
</aside>


<p class="pg">7</p></section>



<section>
<h3>Many box plots, <font class="glow">interactive</font></h3>


<div id="manyboxplots">
</div>

<aside>
Hover over an array above and see the corresponding histogram.  Click
on an array for the histogram to persist; click again to make it go
away.<br><br>

Why aren&rsquo;t all of our graphs interactive?
</aside>


<p class="codelink">
<a href="Coffee/manyboxplots.coffee" class="gray">code</a> |
<a href="Data/hypo.json" class="gray">data</a>


<script type="text/javascript" src="JS/manyboxplots.js"></script>
<p class="pg">8</p></section>





<section>
<h3>Barriers</h3>

<ul class=tight style="margin-right:100px;">
  <li>We never learned how</font></li>
  <li>It&rsquo;s a hassle</font></li>
  <li>No consistent platform</font></li>
  <li>Journal articles are static</font></li>
      <ul class="nodash" style="margin-left:0px;"><li>
          <font class="lowlight"><font class="small">(and what else matters?)</font></font>
      </li></ul>
  <li>Most statisticians are still creating terrible static plots</font></li>
       <ul class="nodash" style="margin-left:0px;"><li>
          <font class="lowlight"><font class="small">(even worse, obnoxious tables)</font></font>
       </li></ul>
</ul>

<aside>
Why aren&rsquo;t statisticians making interactive graphs?  Primarily
because we&rsquo;ve not learned how.  And it does remain somewhat
difficult.  And there are too many options, most of them incompletely
developed.<br><br>

But I think the primary reason is that journal articles are static, so
what incentive is there to develop interactive graphics tools?
<br><br>

And then most statisticians are still putting little care into
visualizations, at all.
</aside>

<p class="pg">9</p></section>

<section>
<h3>Barriers</h3>

<ul class=tight style="margin-right:100px;">
  <li>We never learned how</font></li>
  <li>It&rsquo;s a hassle</font></li>
  <li>No consistent platform</font></li>
  <li>Journal articles are static</font></li>
      <ul class="nodash" style="margin-left:0px;"><li>
          <font class="lowlight"><font class="small">(and what else matters?)</font></font>
      </li></ul>
  <li>Most statisticians are still creating <font class="glow">terrible</font> static plots</font></li>
       <ul class="nodash" style="margin-left:0px;"><li>
       <font class="lowlight"><font class="small">(even worse, <font class="glow">obnoxious</font> tables)</font></font>
       </li></ul>
</ul>

<aside>
I mean it!  Lots of statisticians just create crappy tables, because
they&rsquo;re lazy and don&rsquo;t care.
</aside>

<p class="pg">10</p></section>

<section>
<h3>Examples from <a href="http://amstat.tandfonline.com/loi/jasa">JASA</a></h3>

<div class="chart"><img src="Figs/jasa_sep2012_pg460.png" width=800></div>

<aside>I flipped through the Sept, 2012, issue of JASA.  There are a
surprising number of crappy tables, and a surprising dearth of good
figures.</aside>
<p class="pg">11</p></section>

<section>
<h3>Examples from <a href="http://amstat.tandfonline.com/loi/jasa">JASA</a></h3>

<div class="chart"><img src="Figs/jasa_sep2012_pg600.png" width=800></div>
<p class="pg">12</p></section>

<section>
<h3>Examples from <a href="http://amstat.tandfonline.com/loi/jasa">JASA</a></h3>

<div class="chart"><img src="Figs/jasa_sep2012_pg665.png" width=800></div>
<p class="pg">13</p></section>

<section>
<h3>Examples from <a href="http://amstat.tandfonline.com/loi/jasa">JASA</a></h3>

<div class="chart"><img src="Figs/jasa_sep2012_pg798.png" height=600></div>

<p class="reference">
There <em><font class="highlight">are</font></em> some nice ones.</p>

<aside>Genovese et al., <a href="http://amstat.tandfonline.com/doi/abs/10.1080/01621459.2012.682527">The geometry of nonparametric filament estimation</a>.</aside>
<p class="pg">14</p></section>


<section>
<h3>An alternative to tables</h3>

<div class="chart"><img src="Figs/carr_fig.png"></div>

<p class="reference">
Carr & Pickle (2010) <a
href="http://www.amazon.com/exec/obidos/ASIN/142007573X/7210-20"><i>Visualizing
data patterns with Micromaps</i></a>.</p>

<aside><a href="http://statistics.gmu.edu/people_pages/carr.html">Dan
Carr</a> is brilliant.</aside>

<p class="pg">15</p></section>


<section>
<h3>Options</h3>

<ul class=reallytight>
  <li><a href="http://homepage.cs.uiowa.edu/~luke/xls/xlispstat/">XLisp-Stat</a></li>
  <li><code>locator()</code> and <code>identify()</code></li>
  <li><a href="http://ggobi.org">ggobi</a> and <a href="https://github.com/ggobi/cranvas/wiki">cranvas</a></li>
  <li><a href="http://4dpiecharts.com/2010/10/06/creating-guis-in-r-with-gwidgets/">gWidgets</a></li>
  <li><a href="http://www.stats.gla.ac.uk/~adrian/rpanel/">rpanel</a></li>
  <li><a href="http://rosuda.org/software/Mondrian/">Mondrian</a></li>
  <li><a href="http://rforge.net/Acinonyx/">Acinonyx</a> (aka iPlots eXtreme)</li>
  <li><a href="http://rstudio.org/">RStudio</a>&rsquo;s <code><a
  href="http://rstudio.org/docs/advanced/manipulate">manipulate()</a></code>
  and <a href="http://www.rstudio.com/shiny/">Shiny</a></li>
  <li><a href="http://r-forge.r-project.org/projects/gridsvg/">gridSVG</a></li>
  <li><a href="http://rgl.neoscientists.org">RGL</a></li>
  <li><a href="http://www.youtube.com/watch?v=iSXNfZESR5I">ggplot<font
  class="highlight">3</font></a>?</li>
  <li><a href="http://code.google.com/p/google-motion-charts-with-r/">googleVis</a></li>
  <li><a href="http://d3js.org">D3.js</a></li>
</ul>

<aside>There seem to be <font class="darkhighlight">too many</font> options for producing dynamic and
  interactive graphics.  Many are simple to use but not easily extended.  <a
  href="http://had.co.nz">Hadley Wickham</a> (developer of ggplot2) is
  working on interactive graphics and big data; he might make things easier for
  us.  The most polished interactive graphics are in javascript using
  the D3 library, though for now it&rsquo;s <font class="darkhighlight">very</font>
  low-level work.</aside>

<p class="pg">16</p></section>

<section>
<h3>Trade-offs</h3>

<p align="center" style="margin-top:200px;font-size: 48px;"> <a href="http://code.google.com/p/google-motion-charts-with-r/">Easy</a> <font
  class="gray">or <a href="http://d3js.org" class="highlight">flexible</a>?</font>


<aside>You have to choose between easy or flexible; you can&rsquo;t have
  both.  I&rsquo;m going for flexible.</aside>

<p class="pg">17</p></section>



<section class="nopadding">

<div id="sperm" class="sperm" style="position:fixed;top:-10px;left:-50px;width:1280px;height:960px;">
</div>

<div style="position:fixed;bottom:10px;left:10px;">
<code style="font-size:36px;"><a href="http://bl.ocks.org/1136236">http://bl.ocks.org/1136236</a></code></div>

<aside>
I&rsquo;m focusing on using <a href="http://d3js.org">D3</a>, a javascript
library for data visualization.  (D3 stands for &ldquo;data-driven
documents.&rdquo;)<br><br>

This is one of the images that convinced me that I wanted to learn D3.
The code is surprisingly short and easy to follow.<br><br>

There are lots of additional examples at the <a
href="https://github.com/mbostock/d3/wiki/Gallery">D3 gallery</a>.
</aside>

<p class="pg">18</p>
</div>
<script src="JS/sperm.js"></script>
</section>


<section>
<h3>A lot to learn</h3>

<ul class="small">
  <li><a href="http://d3js.org">D3</a></li>
  <li><a href="https://developer.mozilla.org/en-US/docs/JavaScript">JavaScript</a> /
      <a href="http://coffeescript.org">CoffeeScript</a></li>
  <li><a href="https://developer.mozilla.org/en-US/docs/HTML/HTML5">HTML</a> and &ldquo;<a href="http://en.wikipedia.org/wiki/Document_Object_Model">The DOM</a>&rdquo;</li>
  <li><a href="https://developer.mozilla.org/en-US/docs/CSS">CSS</a></li>
  <li><a href="https://developer.mozilla.org/en-US/docs/SVG">SVG</a></li>
  <li><a href="http://caniuse.com">Differences among web browsers</a></li>
</ul>

<aside>
The hardest part about learning D3 is that there are a whole pile of
things that you need to understand simultaneously, and some of them
are rather dull. <br><br>

You need to learn the D3 library and also to code in JavaScript or
CoffeeScript (a nicer language that is translated to JavaScript).  And
the way D3 works is by manipulating elements in an HTML file,
particularly elements in an SVG object. (SVG stands for
&ldquo;scaleable vector graphics;&rdquo; it&rsquo;s an html-like way to
specify an image.)<br><br>

So you need to know HTML and SVG, and then also CSS (&ldquo;cascading
style sheets&rdquo;), for controlling the style (e.g. color) of
different HTML or SVG elements.<br><br>

Particularly painful is that the different browsers behave somewhat
differently, and you need to take that into account if you want your
visualizations to be useable in any browser.
</aside>

<p class="pg">19</p></section>
</section>

<section>
<h3>But it&rsquo;s getting simpler</h3>

<ul class="small">
  <li><a href="https://github.com/nachocab/clickme">clickme</a>
  package for R</li>
  <li><a href="http://healthvis.org">healthvis</a> for R</li>
  <li><a href="https://github.com/trifacta/vega">Vega</a> (more
  high-level)</li>
  <li><a href="https://github.com/PatMartin/DexCharts">DexCharts</a>
  (reuseable charts for D3)</li>
</ul>

<aside>
There are a lot of people working to make these things easier, or
accessible from R, or both.
</aside>

<p class="pg">20</p></section>
</section>



<section>
<h3>Others&rsquo; <a href="http://d3js.org" style="color:white;">D3</a> <a href="https://github.com/mbostock/d3/wiki/Gallery" class="highlight">examples</a></h3>

<ul class=tight>
  <li><a href="http://bost.ocks.org/mike/nations/">Wealth and health of nations</a></li>
  <li><a href="http://bl.ocks.org/1667367">Brushing</a></li>
  <li><a href="http://bost.ocks.org/mike/miserables/">Adjacency matrix</a></li>
  <li><a href="http://bost.ocks.org/mike/uberdata/">Chord diagram</a></li>
  <li><a href="http://bost.ocks.org/mike/shuffle/">Random shuffles</a></li>
  <li><a href="http://mbostock.github.com/d3/talk/20111018/collision.html">Collisions</a></li>
  <li><a href="http://mbostock.github.com/d3/talk/20110921/#6">Different kinds of forces</a></li>
  <li><a href="http://www.nytimes.com/interactive/2012/09/06/us/politics/convention-word-counts.html">Word counts at the conventions</a></li>
</ul>

<aside>There are loads of cool examples on the web.</aside>

<p class="pg">21</p></section>

</section>



<section>
<h3>Interactive LOD curves</h3>

<div id="lod_and_effect" class="fig"></div>


<aside>
The curves at the top correspond to a measure of association between
genotype and insulin level.  LOD scores are log<sub>10</sub>
likelihood ratios: a transformation of the F statistic from
ANOVA.<br><br>

Click on a chromosome in the top panel to see a detailed view in the
lower-left panel.  Click on a marker in the lower-left panel to see
effect estimates and raw data on the right.  Hover over things on the
right to get some further information.
</aside>


<p class="codelink">
<a href="Coffee/lod_and_effect.coffee" class="gray">code</a> |
<a href="Data/insulinlod.json" class="gray">data</a>


<script type="text/javascript" src="JS/lod_and_effect.js"></script>
<p class="pg">22</p></section>


<section nopadding>

<h3 style="margin-top:10px;margin-left:-20px;">Interactive eQTL plot</h3>

<div id="cistrans" class="fig" style="margin-top:-80px;"></div>


<aside>
The top-left image shows inferred eQTL (sites in genome where genotype
is associated with the expression of some gene).  The y-axis
corresponds to the genomic location of probes on a gene expression
array.  The x-axis corresponds to marker positions.  Dots indicate
that the genotypes at a particular marker are associated with the
expression of a particular probe.<br><br>

Hover over a point in the top-right to see all eQTL for a given probe.
Click on a point to see the LOD curves across the genome for that
probe: the association between that probe&rsquo;s expression and genotype at
each position in the genome.<br><br>

Click on a marker in the lower LOD curve plot to see the corresponding
phenotype-vs-genotype plot in the top-right.
</aside>


<p class="codelink">
<a href="Coffee/cistrans.coffee" class="gray">code</a>


<script type="text/javascript" src="JS/cistrans.js"></script>

<p class="pg">23</p></section>


<section>

<h3>Spalding project</h3>

<div style="position:absolute;top:312px;left:420px;font-size:20px;">
<a href="Movie/gmovie.mov">
  <img src="Movie/initial_view.png" width="440" alt="Phenotype movie"/>
</a></div>

<aside>
Click on the image to see a movie.<br><br>

You&rsquo;ll be popped to a separate page; you&rsquo;ll need to navigate back.
</aside>

<p class="pg">24</p></section>

</section>



<section nopadding>

<h3 style="margin-top:10px;margin-left:-20px;">Phenotypes measured over time</h3>

<div id="phenofig" style="margin-top:-50px;margin-left:100px;"></div>

<aside>
The top panel is an image plot of the phenotype, measured over time,
for many lines.  The rows correspond to lines.
There is a color scale at the bottom right.<br><br>

Hover over a row to see the detailed time course below.<br><br>

Click on a row and the curve below will be retained as you move
away.  Click again (if you can get your pointer onto it) and it will
be removed.
</aside>


<p class="codelink">
<a href="Coffee/pheno.coffee" class="gray">code</a> |
<a href="Data/pheno.json" class="gray">data</a>


<script type="text/javascript" src="JS/pheno.js"></script>
<p class="pg">25</p></section>



<section>

<h3>LOD curves at 62 min</h3>

<div id="lod_onetime_fig" style="margin-left:-30px;"></div>

<aside>
LOD curves and QTL effect plot for a single time point.

Click on markers to see the phenotype-genotype relationship to the
right.
</aside>


<p class="codelink">
<a href="Coffee/lod_onetime.coffee" class="gray">code</a> |
<a href="Data/ontime.json" class="gray">data</a>


<script type="text/javascript" src="JS/lod_onetime.js"></script>
<p class="pg">26</p></section>


<section>

<h3>LOD curves at 62 min</h3>

<div id="lod_onetime_random_fig" style="margin-left:-30px;"></div>

<aside>
Illustration of permutation test: Click the green button to shuffle
the phenotypes and re-draw the LOD curves.<br><br>

There&rsquo;s a hidden &ldquo;back&rdquo; button a bit to the right of
the green button.
</aside>

<p class="codelink">
<a href="Coffee/lod_onetime_random.coffee" class="gray">code</a> |
<a href="Data/ontime_random.json" class="gray">data</a>


<script type="text/javascript" src="JS/lod_onetime_random.js"></script>
<p class="pg">27</p></section>


<section nopadding>

<h3>All times, considered individually</h3>


<div id="lod_alltimes_fig" style="font-size:20px;margin-left:-50px;"></div>

<aside>
Here we consider a phenotype measured over time, and perform QTL
analysis with each time point individually, to seek genetic loci that
affect the response.  The data concern a set of 162 recombinant inbred
lines.<br><br>

The top-left plot is a &ldquo;heat map&rdquo; of the LOD scores
for each time point at each genomic position.  LOD scores are also
colored to indicate the sign of the QTL effect, with red indicating
that BB lines have larger phenotype values and blue indicating that AA lines
have larger phenotype values.  We consider only those (position, time)
pairs with LOD > 1.<br><br>

When you hover over a point in the top-left plot, the LOD curves for the
corresponding time are shown below, and the phenotype averages and
estimated QTL effect (across time) for the corresponding genomic
position are shown to the right.<br><br>

Click on a point to show pointwise confidence bands on the QTL
effect (&plusmn; 2 SE).  (We require a mouse click, as otherwise the graph
was painfully slow to refresh.)
</aside>


<p class="codelink">
<a href="Coffee/lod_alltimes.coffee" class="gray">code</a> |
<a href="Data/alltimes.json" class="gray">data</a> |
<a href="alt_alltimes.html" class="gray">separate view</a>

<script type="text/javascript" src="JS/lod_alltimes.js"></script>
<p class="pg">28</p></section>



<section>
<h3>Summary</h3>

<ul class="tight">
  <li>For high-dimensional data, good visualizations are <font
  class="highlight">critical</font></li>

  <li><font class="glow">Interactive</font> graphics
  require effort, but they</li>

      <ul class="lowlight">
        <li>Facilitate exploration</li>
        <li>Are great collaborative tools</li>
        <li>Enable summaries with access to the details</li>
      </ul>

  <li>Visualizations must be <font
  class="highlight">tailored</font> to the data and
  questions/goals

  <li><a href="http://d3js.org" class="highlight">D3</a> (aka <a
  href="http://www.d3js.org" class="highlight">d3.js</a>) is rather low level, but it</li>
      <ul class="lowlight">
        <li>Is totally flexible (like <a href="http://www.r-project.org">R</a>&rsquos static graphics)</li>
        <li>Provides hours of enjoyment</li>
        <li>Can provide other hours of frustration</li>
      </ul>

</ul>

<aside>
It&rsquo;s always good to have a summary slide with your key points.
</aside>

<p class="pg">29</p></section>




<section>
<h3>Acknowledgments<font class="lowlight">: Data</font></h3>

<div style="margin-left:120px;font-size:28px;">

<table cellspacing=25>
  <tr valign="top">
     <td class="lowlight">Alan Attie<br>Mark Keller</td>
     <td width=10></td>
     <td>Biochemistry, UW&ndash;Madison</td>
     </tr>

  <tr valign="top">
     <td class="lowlight">Brian Yandell</td>
     <td></td>
     <td>Statistics and Horticulture, UW&ndash;Madison</td>
     </tr>

  <tr valign="top">
     <td class="lowlight">Christina Kendziorski<br>Aimee Teo Broman</td>
     <td></td>
     <td>Biostatistics & Medical Informatics, UW&ndash;Madison</td>
     </tr>

  <tr valign="top">
     <td class="lowlight">Eric Schadt
     <td></td>
     <td>Pacific Biosciences of California</td>
     </tr>

  <tr valign="top">
     <td class="lowlight">Danielle Greenawalt<br>Amit Kulkarni</td>
     <td></td>
     <td>Merck & Co., Inc.</td>
     </tr>

  <tr height=30><td></td></tr>

  <tr valign="top">
     <td class="lowlight">Edgar Spalding<br>Candace Moore<br>Logan Johnson</td>
     <td></td>
     <td>Botany, UW-Madison</td>
     </tr>
</table>
</div>

<aside>
Many people are involved in the studies that led to this work.
</aside>

<p class="pg">30</p></section>

<section>
<h3>Acknowledgments<font class="lowlight">: Code</font></h3>

<div style="margin-left:120px">
  <p><a href="http://d3js.org" class="highlight">D3</a> and <a href="https://github.com/mbostock/stack" class="highlight">stack.js</a> by <a href="http://bost.ocks.org/mike" rel="author">Mike Bostock</a>

  <p><a href="http://alignedleft.com/tutorials/d3"
  class="highlight">D3 tutorials</a> by <a href="http://alignedleft.com">Scott Murray</a>

  <p><a href="http://coffeescript.org" class="highlight">CoffeeScript</a> by <a
  href="https://github.com/jashkenas">Jeremy Ashkenas</a>

  <p><a href="http://eloquentjavascript.net"
  class="highlight"><i>Eloquent JavaScript</i></a> by <a href="http://marijnhaverbeke.nl">Marijn Haverbeke</a>

  <p><a href="http://pragprog.com/book/tbcoffee/coffeescript"
  class="highlight"><i>CoffeeScript</i> book</a> by <a href="http://trevorburnham.com">Trevor Burnham</a>

  <p><a href="https://github.com/Caged/d3-tip" class="highlight">d3-tip</a> by <a href="http://labratrevenge.com">Justin Palmer</a>
</div>

<aside>
This is the last slide.<br><br>

The interactive examples
were produced with the <a href="http://d3js.org">D3</a>
library (written in JavaScript). I wrote the code in CoffeeScript,
which is translated to JavaScript but is a better-designed language.  <a
href="http://eloquentjavascript.net"><i>Eloquent JavaScript</i></a> is
a great free book.  The <a
href="http://pragprog.com/book/tbcoffee/coffeescript"><i>CoffeeScript</i></a>
book is not free but is also great; you need to learn at least a bit of
JavaScript first.  <a href="http://alignedleft.com/tutorials/d3">Scott
Murray&rsquo;s D3 tutorials</a> are superb and are a great place to start
for learning D3.  These slides are in html, using <a
href="https://github.com/mbostock/stack">stack.js</a>.  I used <a
href="http://github.com/kbroman/d3-tip">a modified version</a> of <a
href="http://github.com/Caged/d3-tip">d3-tip</a> for tool tips.
</aside>

<p class="pg">31</p></section>

<script src="JS/stack.v0.js"></script>
</body>
</html>